modules/images/image-loading-optimization/optimization.php

```diff
@@ -0,0 +1,174 @@
+		$link_tag = '<link data-ilo-added-tag rel="preload" fetchpriority="high" as="image"';
+		foreach ( array_filter( $img_attributes ) as $name => $value ) {
+			// Map img attribute name to link attribute name.
+			if ( 'srcset' === $name || 'sizes' === $name ) {
+				$name = 'image' . $name;
+			} elseif ( 'src' === $name ) {
+				$name = 'href';
+			}
+
+			$link_tag .= sprintf( ' %s="%s"', $name, esc_attr( $value ) );
+		}
+		$link_tag .= ">\n";
+
+		$preload_links[] = $link_tag;
+	}
+
+	return implode( '', $preload_links );
+}
+
+/**
+ * Optimizes template output buffer.
+ *
+ * @since n.e.x.t
+ * @access private
+ *
+ * @param string $buffer Template output buffer.
+ * @return string Filtered template output buffer.
+ */
+function ilo_optimize_template_output_buffer( string $buffer ): string {
+	$slug = ilo_get_url_metrics_slug( ilo_get_normalized_query_vars() );
+	$post = ilo_get_url_metrics_post( $slug );
+
+	// No URL metrics are present, so there's nothing we can do.
+	if ( ! $post ) {
+		return $buffer;
+	}
+
+	$url_metrics = ilo_parse_stored_url_metrics( $post );
+
+	$breakpoint_max_widths                   = ilo_get_breakpoint_max_widths();
+	$url_metrics_grouped_by_breakpoint       = ilo_group_url_metrics_by_breakpoint( $url_metrics, $breakpoint_max_widths );
+	$lcp_elements_by_minimum_viewport_widths = ilo_get_lcp_elements_by_minimum_viewport_widths( $url_metrics_grouped_by_breakpoint );
+	$all_breakpoints_have_url_metrics        = count( array_filter( $url_metrics_grouped_by_breakpoint ) ) === count( $breakpoint_max_widths ) + 1;
+
+	// TODO: Handle case when the LCP element is not an image at all, but rather a background-image.
+	// Prepare to set fetchpriority attribute on the image when all breakpoints have the same LCP element.
+	if (
+		// All breakpoints share the same LCP element (or all have none at all).
+		1 === count( $lcp_elements_by_minimum_viewport_widths )
+		&&
+		// The breakpoints don't share a common lack of an LCP element.
+		! in_array( false, $lcp_elements_by_minimum_viewport_widths, true )
+		&&
+		// All breakpoints have URL metrics being reported.
+		$all_breakpoints_have_url_metrics
+	) {
+		$common_lcp_element = current( $lcp_elements_by_minimum_viewport_widths );
+	} else {
+		$common_lcp_element = null;
+	}
+
+	// Walk over all IMG tags in the document and ensure fetchpriority is set/removed, and gather IMG attributes for preloading.
+	$processor = new ILO_HTML_Tag_Processor( $buffer );
+	foreach ( $processor->open_tags() as $tag_name ) {
+		if ( 'IMG' !== $tag_name ) {
+			continue;
+		}
+
+		// Ensure the fetchpriority attribute is set on the element properly.
+		if ( $common_lcp_element && $processor->get_breadcrumbs() === $common_lcp_element['breadcrumbs'] ) {
+			if ( 'high' === $processor->get_attribute( 'fetchpriority' ) ) {
+				$processor->set_attribute( 'data-ilo-fetchpriority-already-added', true );
+			} else {
+				$processor->set_attribute( 'fetchpriority', 'high' );
```

Comment:
Shouldn't we also remove a potential `loading="lazy"` attribute here? Or is this completely irrelevant in that case and `fetchpriority="high"` takes precendence?

If the latter, I think that would be good to document in the code somewhere, e.g. in this function's doc block.

Reply:
You're right. It would be removed here. But I was intending to leave processing of lazy-loading optimizations to a subsequent PR. I'll add a TODO here for that.

```suggestion
				// TODO: When optimizing lazy-loading, this should also remove any `loading` attribute here.
				$processor->set_attribute( 'fetchpriority', 'high' );
```

Reply:
I went ahead and made this change: 9d8261535314ef0bfe79ba2294ab0150fa54f764.

Nevertheless, since the LCP images also get `preload` links added for them, the presence of `loading=lazy` actually won't hurt performance for the LCP image. This is based on my findings in https://github.com/WordPress/performance/issues/117#issuecomment-1773456772.

Conversely, in the `else` condition here, we actually need to _add_ `loading=lazy` to breakpoint-specific LCP images if they aren't visible on all breakpoints. This I'll leave for another PR.